src/main/java/org/apache/sling/pipes/internal/ACLPipe.java

```diff
@@ -0,0 +1,260 @@
+     * @param resource configuration resource
+     * @throws Exception bad configuration handling
+     */
+    public ACLPipe(Plumber plumber, Resource resource, PipeBindings upperBindings) throws Exception {
+        super(plumber, resource, upperBindings);
+        resolver = resource.getResourceResolver();
+        session = resolver.adaptTo(Session.class);
+        userManager = resolver.adaptTo(UserManager.class);
+        if (getConfiguration() != null) {
+            ValueMap properties = getConfiguration().adaptTo(ValueMap.class);
+            privilegesInput = properties.get(JCR_PRIVILEGES_INPUT, new String[]{});
+            allow = properties.get(PN_ALLOW, false);
+            deny = properties.get(PN_DENY, false);
+            authorizable = properties.get(PN_AUTHORIZABLE, false);
+        }
+    }
+
+    @Override
+    public Iterator<Resource> computeOutput() throws Exception {
+        Resource resource = getInput();
+        if (resource != null) {
+            if (allow || deny) {
+                logger.info("Going to changing ACL for the resource at path {}", resource.getPath());
+                if (StringUtils.isEmpty(getExpr())) {
+                    throw new Exception("expression for the principal or authorizable Id should be provided or provided correctly for privileges to be set");
+                }
+                Principal principal = getPrincipalFor(getExpr());
+                if ( null != privilegesInput && privilegesInput.length == 0 ){
+                    // create a privilege set with jcr:all
+                    privileges = allow ? AccessControlUtils.privilegesFromNames(session, Privilege.JCR_ALL) : AccessControlUtils.privilegesFromNames(session, Privilege.JCR_READ);
+                }
+                else {
+                    try {
+                        privilegesInput = processPrivilegesInput(privilegesInput);
+                        privileges = AccessControlUtils.privilegesFromNames(session, privilegesInput);
+                    }
+                    catch (Exception e) {
+                        logger.error("unable to get privileges , either privileges input is wrong either error evaluting bindings for privileges", e);
+                    }
+                }
+                if (!isDryRun()) {
+                    if (allow) {
+                        logger.info("adding privileges {} for principal {} allow {}",ArrayUtils.toString(privileges), principal.getName(), allow);
+                        AccessControlUtils.addAccessControlEntry(session, resource.getPath(), principal, privileges, true);
+                    } else if (deny) {
+                        logger.info("adding privileges {} for principal {} deny {}",ArrayUtils.toString(privileges), principal.getName(), deny);
+                        AccessControlUtils.addAccessControlEntry(session, resource.getPath(), principal, privileges, false);
+                    }
+                    if (session.hasPendingChanges()) {
```

Comment:
if you can, please not save here, plumber will do that for you and saving less will be more performant